keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -1880,82 +1911,240 @@ progress.
+in the real cluster, it can determine what resource the vendor driver would
+be able to make available if it was created for real.
+
+#### Generic plugin enhancements
+
+The changes in this section are independent of DRA. They could also be used to
+simulate volume provisioning better.
 
 At the start of a scale up or scale down cycle, autoscaler takes a snapshot of
 the current cluster state. Then autoscaler determines whether a real or
 fictional node fits a pod by calling the pre-filter and filter extension points
 of scheduler plugins. If a pod fits a node, the snapshot is updated by calling
 [NodeInfo.AddPod](https://github.com/kubernetes/kubernetes/blob/7e3c98fd303359cb9f79dfc691da733a6ca2a6e3/pkg/scheduler/framework/types.go#L620-L623). This
-influences further checks for other pending pods.
+influences further checks for other pending pods. During scale down, eviction
+is simulated by
+[SimulateNodeRemoval](https://github.com/kubernetes/autoscaler/blob/2f7c61e13bd1cbfc0ba4085fb84bd692a1e9ac6e/cluster-autoscaler/simulator/cluster.go#L149)
+which [pretends that pods running on a node that is to be removed are not
+running](https://github.com/kubernetes/autoscaler/blob/2f7c61e13bd1cbfc0ba4085fb84bd692a1e9ac6e/cluster-autoscaler/simulator/cluster.go#L231-L237).
 
-To support the custom allocation logic that a vendor uses for its resources,
-the autoscaler needs an extension mechanism similar to the [scheduler
-extender](https://github.com/kubernetes/kubernetes/blob/7e3c98fd303359cb9f79dfc691da733a6ca2a6e3/pkg/scheduler/framework/extender.go#L24-L72). The
-existing scheduler extender API has to be extended to include methods that
-would only get called by the autoscaler, like starting a cycle. Instead of
-adding these methods to the scheduler framework, autoscaler can define its own
-interface that inherits from the framework:
+The DRA scheduler plugin gets integrated into this snapshotting and simulated
+pod scheduling through a new scheduler framework interface:
 
 ```
-import "k8s.io/pkg/scheduler/framework"
-
-type Extender interface {
-    framework.Extender
-
-    // NodeSelected gets called when the autoscaler determined that
-    // a pod should run on a node.
-    NodeSelected(pod *v1.Pod, node *v1.Node) error
-
-    // NodeReady gets called by the autoscaler to check whether
-    // a new node is fully initialized.
-    NodeReady(nodeName string) (bool, error)
+// ClusterAutoScalerPlugin is an interface that is used only by the cluster autoscaler.
+// It enables plugins to store state across different scheduling cycles.
+//
+// The usual call sequence of a plugin when used in the scheduler is:
+// - at program startup:
+//   - instantiate plugin
+//   - EventsToRegister
+// - for each new pod:
+//   - PreEnqueue
+// - for each pod that is ready to be scheduled, one pod at a time:
+//   - PreFilter, Filter, etc.
+//
+// Cluster autoscaler works a bit differently. It identifies all pending pods,
```

Comment:
What about Karpenter?  It's got a lot of attention lately, and I think we need to consider how it would engage with this feature.  There are others, but I don't hear much about them - we may need to do outreach.

Reply:
@ellistarn

AFAIK Karpenter doesn't import scheduler plugins the way CAS does. Instead it re-implements the underlying logic in its own framework (keep me honest here @ellistarn), which would require yet another re-implementation of the plugin logic? Maybe an RPC call would be more feasible in Karpenter than it is in CAS?

Reply:
Yeah we don't currently support scheduler plugins. Let me read up a bit and form an opinion on this one.

Reply:
I'm not actually that concerned about the scheduler implementation. It's a much longer discussion, but due to the higher dimensionality of karpenter than CAS, we can't use the schedulers "search" approach. 

I am a bit sad that this KEP overhauls extended resources, but fails to address a critical issue that NodeAutoscalers face with node capacity discovery. CAS faces this with scale to zero, where a users must manually tag their asgs (or an equivalent vendor specific analogue) with the extended resources that will be attached to the node. 

Karpenter faces a similar problem, where we can't know that a node will have an extended resource until a node launches. The AWS cloud provider currently hardcodes knowledge of a small set of extended resources (Nvidia, etc) that are most relevant to our customers, but similar to CAS discovery, we will need to build a discovery mechanism to support 3rd party plugins.

Reply:
As prevalent as autoscaling is, I think it is safe to say that this feature is open to new ideas, up to and including "throw the whole thing away and start over", if needed.